### ISSUE TITLE
Installer

### ISSUE THREAD

**Issue by Mokou** [thpatch/thcrap#31]
We need one. I choose NSIS.

**Comment by Tudi20** [comment]
Configurator and Installer became the same thing, for configure see #8 .

**Comment by Mokou** [comment]
I disagree. #8 doesn’t solve this since the GUI still needs to be installed. Distributing a .zip is not a good experience.

**Comment by Tudi20** [comment]
Self-extracting zip, better?

**Comment by Tudi20** [comment]
And also 99% of times, people get thcrap bundled with a Touhou game.

**Comment by ManDude** [comment]
Those people only use that thcrap for that Touhou game and no other.

**Comment by Tudi20** [comment]
### Options
#### Windows
- Self-extracting zip (this is a .exe file)
- An MSIX mindless click-click-click installer
- Our own .exe installer

Furthermore:
Any .exe, .msi, or .msix app could go to Windows Package Manager (`winget`) that adheres to [these policies](https://docs.microsoft.com/en-us/windows/package-manager/package/windows-package-manager-policies).
If we package as msix with [this](https://docs.microsoft.com/en-us/windows/msix/desktop/source-code-overview) it will become and uwp app, then we could distribute to Windows Store, if we adhere to [these policies](https://docs.microsoft.com/en-us/windows/uwp/publish/store-policies).
#### MacOS
Somebody who uses MacOS please fill us in, but I have a suspecting feeling we are definitely not going to make it to MacOs's Store. But maybe the Linux methods would work here as well.
#### Linux
I think the best idea is to create .AppImage, since that basically runs everywhere, and then either provide an auto-updater for it somehow, or also maintain a Flatpack for Debian-based, and on the AUR for Arch-based. (The rest can figure it out themselves.)

**Comment by Tudi20** [comment]
> Almost all of the installers on Windows except ones that install huge programs are actually self extracting archives. In fact, if you have 7-zip installed in Windows, you can just right click on the installer and it will show you to open in 7-zip. This is because for an executable in the PE file format, there is a specific last byte, after which it is all done and Windows loader will not load it into memory.
This trick was invented my Microsoft to my knowledge, there used to be a program called iexpress baked into Windows since Windows 95 and it existed in Windows 10 too. Not sure if it is in 11 though as I have switched completely to Linux. The iexpress tool allowed you to create these self extracting archives, and it used the CAB archive format. Then most common after that was WinZIP SFX and WinRAR SFX archives and later 7zip SFX. Very rarely, there are Uharc SFX too, made by WinUharc and Uharc CMD projects.
Most archive programs will try to look for the starting signature in input bytestream before throwing an error, where this is usually 2-4 dedicated bytes which will tell the file format and the format version. So all the SFX executable needs to do is to get its own file name (in C/C++ it is the 0th argument in argv array) and pass it to the archiving program. Sometimes instead of an archiving program, these link with the libraries so unarchiving code is also in the same EXE file.
Btw, the same is true with MSI installers too. MSI is just Microsoft's format which say how an installer should look like and behave. InstallShield uses multiple files though, and they are kept in different .DAT or .BIN files. InnoSetup allows both options - if SFX route, it uses LZMA compression. Else .BIN files are used as archives. NSIS, allows the package script author to decide the compression format. Most will use LZMA but other formats are available too.
However, these are for older installers or for installers for third party software. Most used softwares now go for Web Installers, which downloads the real archive in the background and install it over the network. This is also done for telemetry reasons. Installer can send over the user-agent which contains OS version and also system specifications to the server which is used to know the reach and analytics. - [Youtube comment by Sri Harsha Chilakapati](https://www.youtube.com/watch?v=bwTEAj5eF6o&lc=Ugy32AOKeepiPpFAOsx4AaABAg)

**Comment by Tudi20** [comment]
MSIX Docs: https://learn.microsoft.com/en-us/windows/msix/packaging-tool/create-an-msix-overview
Alternative: https://github.com/Squirrel/Squirrel.Windows
Our Own: Alternatively, we could reuse the existing thcrap-updater, and make it into a web installer.

**Comment by Tudi20** [comment]
Even more alternatively, we could distribute via https://github.com/ScoopInstaller/Scoop

**Comment by lilyremigia** [comment]
To do:

- [ ] Add installation logic to configure
- [ ] Add uninstallation logic to configure
- [ ] Update deployment script
- [ ] Optional: Update nightly deployment script

**Comment by brliron** [comment]
Ideally, we'd also support everything required to install thcrap in a write-protected directory (program files) and running as a normal user:
- Self-elevation during self updates (asking for admin rights when we detect that a new update is available (but only if the thcrap directory isn't writable), and failing gracefully if our request is denied).
- Being able to use AppData for the config, logs and repos. I guess a good way to handle that would be with a config file in <thcrap_dir>/config with all the paths (all of them defaulting to the current directory if this file or something in it is missing), and having the installer create a default one pointing to AppData if thcrap is installed in Program Files (I guess we can have a checkbox "install in portable mode" or something like that somewhere to make that choice).
- Also, we probably want to hide the "create shortcuts in the thcrap folder" checkbox in configure_v3 if the thcrap folder isn't writable.
